File: java/src/array/BinaryTree.java
package array;

public class BinaryTree {

    private TreeNode root;

    private StringBuffer sb;

    public void setRoot(TreeNode node) {
        this.root = node;
    }

    public String depthFirstPreOrderTraverse() {
        sb = new StringBuffer();
        dfsPreOrder(this.root);

        return sb.toString();
    }

    private void dfsPreOrder(TreeNode node) {
        if (node == null)
            return;

        sb.append(node.getKey());
        sb.append(", ");

        dfsPreOrder(node.getLeft());
        dfsPreOrder(node.getRight());
    }

    public String depthFirstInOrderTraverse() {
        sb = new StringBuffer();
        dfsInOrder(this.root);

        return sb.toString();
    }

    private void dfsInOrder(TreeNode node) {
        if (node == null)
            return;

        dfsInOrder(node.getLeft());

        sb.append(node.getKey());
        sb.append(", ");

        dfsInOrder(node.getRight());
    }

    public String depthFirstPostOrderTraverse() {
        sb = new StringBuffer();
        dfsPostOrder(this.root);

        return sb.toString();
    }

    private void dfsPostOrder(TreeNode node) {
        if (node == null)
            return;

        dfsPostOrder(node.getLeft());
        dfsPostOrder(node.getRight());

        sb.append(node.getKey());
        sb.append(", ");
    }

    private class Queue<E> {
        private Node head;

        private Node tail;

        private class Node {
            public E key;

            public Node next;

            public Node(E key) {
                this.key = key;
            }
        }

        public void enqueue(E key) {
            Node node = new Node(key);

            if (head == null)
                head = node;

            if (tail != null)
                tail.next = node;

            tail = node;
        }

        public E dequeue() {
            if (head == null)
                return null;

            Node del = head;
            E returnValue = del.key;
            head = del.next;
            del = null;

            return returnValue;
        }

        public boolean empty() {
            return head == null;
        }
    }

    public String breathFirst() {
        sb = new StringBuffer();

        Queue<TreeNode> queue = new Queue<TreeNode>();
        queue.enqueue(this.root);

        while(!queue.empty()) {
            TreeNode<String> node = queue.dequeue();
            sb.append(node.getKey());
            sb.append(", ");

            if (node.getLeft() != null)
                queue.enqueue(node.getLeft());

            if (node.getRight() != null)
                queue.enqueue(node.getRight());
        }

        return sb.toString();
    }
}

File: python/array/binaryTree.py
class BinaryTree:
    def __init__(self, root):
        self.root = root
        self.result = []

    def depthFirstPreOrderTraverse(self):
        self.__preDFT(self.root)
        return ','.join(self.result)

    def __preDFT(self, node):
        if node == None:
            return
        self.result.append(node.data)
        self.__preDFT(node.left)
        self.__preDFT(node.right)

    def depthFirstInOrderTraverse(self):
        self.__inDFT(self.root)
        return ','.join(self.result)

    def __inDFT(self, node):
        if node == None:
            return
        self.__inDFT(node.left)
        self.result.append(node.data)
        self.__inDFT(node.right)

    def depthFirstPostOrderTraverse(self):
        self.__postDFT(self.root)
        return ','.join(self.result)

    def __postDFT(self, node):
        if node == None:
            return
        self.__postDFT(node.left)
        self.__postDFT(node.right)
        self.result.append(node.data)

    class Queue:
        head = None
        tail = None

        class Node:
            def __init__(self, data):
                self.data = data
                self.next = None
        
        def enqueue(self, input):
            node = self.Node(input)

            if self.head is None:
                self.head = node

            if self.tail is not None:
                self.tail.next = node
            
            self.tail = node

        def dequeue(self):
            if self.head is None:
                return None

            delete = self.head
            self.head = delete.next
            result = delete.data
            delete = None

            return result

        def empty(self):
            return self.head is None

    def breathFirst(self):
        queue = self.Queue()
        queue.enqueue(self.root)

        while queue.empty() == False:
            treeNode = queue.dequeue()
            self.result.append(treeNode.data)
            if treeNode.left is not None:
                queue.enqueue(treeNode.left)
            if treeNode.right is not None:
                queue.enqueue(treeNode.right)
        
        return ','.join(self.result)


